src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,134 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final Object[] DEFAULTCAPACITY_EMPTY_ELEMENTDATA = {};
+    private Object[] elementData;
+    private int size;
+
+    public ArrayList() {
+        this.elementData = DEFAULTCAPACITY_EMPTY_ELEMENTDATA;
+    }
+
+    private Object[] grow(int minCapacity) {
+        int oldCapacity = elementData.length;
+        Object[] oldElement = elementData;
+        if (oldCapacity > 0 || elementData != DEFAULTCAPACITY_EMPTY_ELEMENTDATA) {
+            int prefLength = oldCapacity + Math.max(minCapacity - oldCapacity, oldCapacity >> 1);
```

Comment:
line 19 👍

Reply:
Math.max(minCapacity - oldCapacity, oldCapacity >> 1) I've actualy looked it in default ArrayList. But I thought it's a good logic for addAll. If adding array is bigger than half of our elementData, we'll need to grov twice. And it's will last too long, so better to grow once to the right size.